internal/jptrace/sanitizer/emptyservicename.go

```diff
@@ -23,12 +23,44 @@ func NewEmptyServiceNameSanitizer() Func {
 }
 
 func sanitizeEmptyServiceName(traces ptrace.Traces) ptrace.Traces {
+	needsModification := false
 	resourceSpans := traces.ResourceSpans()
 	for i := 0; i < resourceSpans.Len(); i++ {
 		resourceSpan := resourceSpans.At(i)
 		attributes := resourceSpan.Resource().Attributes()
 		serviceName, ok := attributes.Get(string(otelsemconv.ServiceNameKey))
 		switch {
+		case !ok:
+			needsModification = true
+		case serviceName.Type() != pcommon.ValueTypeStr:
+			needsModification = true
+		case serviceName.Str() == "":
+			needsModification = true
+		}
+		if needsModification {
+			break
+		}
+	}
+
+	if !needsModification {
+		return traces
+	}
+
+	var workingTraces ptrace.Traces
+
+	if traces.IsReadOnly() {
+		workingTraces = ptrace.NewTraces()
+		traces.CopyTo(workingTraces)
+	} else {
+		workingTraces = traces
+	}
+
+	workingResourceSpans := workingTraces.ResourceSpans()
+	for i := 0; i < workingResourceSpans.Len(); i++ {
+		resourceSpan := workingResourceSpans.At(i)
```

Comment:
I think you can simplify using iterators
```suggestion
	for resourceSpan := range workingTraces.ResourceSpans().All()  {
```